016-database-adapters.md

```diff
@@ -0,0 +1,340 @@
+
+**What is the context or background in which this problem exists?**
+
+Larger data set and more complex projects, require more flexibility and transparency regarding the database inner structure and performance capabilities.
+
+**Once the proposal is implemented, how will the system change?**
+
+We will implement simplified data structures that could be manipulated from both the Appwrite dashboard and directly from the relevant connected DB. We will also add new endpoints to easily add, delete, and view collection rules and indexes.
+
+<!-- Write your answer below. -->
+
+<!-- Please avoid discussing your proposed solution. -->
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail keeping in mind the following:
+
+- Its interaction with other parts of the system is clear
+- It is reasonably clear how the contribution would be implemented
+- Dependencies on libraries, tools, projects or work that isn't yet complete
+- New API routes that need to be created or modifications to the existing routes (if needed)
+- Any breaking changes and ways in which we can ensure backward compatibility.
+- Use Cases
+- Goals
+- Deliverables
+- Changes to documentation
+- Ways to scale the solution
+
+Ensure that you include examples, code-snippets etc. to allow the community to understand the proposed solution. **It would be best if the examples use naming conventions that you intend to use during the actual implementation so that changes can be suggested early on during the development.**
+
+Write your answer below.
+
+-->
+
+### Utopia Database
+
+We'll implement an extra database abstraction layer as part of the Utopia project. This layer will include support for multiple database, and will expose a consistent API which will act as the lowest common denominator for all the supported adapters.
+
+The list of supported adapters will include:
+* Postgres
+* MySQL
+* MariaDB
+* MongoDB¹
+
+#### Data Types
+
+This library will support storing and fetching of all common JSON simple and complex [data types](https://restfulapi.net/json-data-types/).
+
+##### Simple Types
+
+* String
+* Integer
+* Float
+* Boolean
+* Null
+
+##### Complex Types
+* Array
+* Object
+* Relationships
+  * Reference (collection / document)
+  * References (Array of - collection / document)
+
+> Databases that don't support the storage of complex data types should store them as strings and parse them correctly when fetched.
+
+#### Persistency
+
+Each database adapter should support the following actions for fast storing and retrieval of collections of documents.
+
+**Databases** (Schemas for SQL)
+* create
+* delete
+
+**Collections** (Tables for SQL)
+* createCollection($name)
```

Comment:
How should the adapter work with the createCollection API route?
In the PoC, [createCollection](https://github.com/utopia-php/database/blob/v0/src/Database/Database.php#L146-L156) returns a bool, but in the createCollection API route, [createDocument](https://github.com/appwrite/appwrite/blob/master/app/controllers/api/database.php#L58) is used, and a document is returned. I'm not sure how the new createCollection route should look - should the adapter have a `getCollection` method?

Reply:
I think we have 2 options, either have a get collection, or we can store the collection metadata as a document. Although the names are the same, this are 2 separate layers that don't have to act exactly the same.